streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -82,6 +82,37 @@ abstract class ZStream[-R, +E, +O](
   ): ZStream[R1, E1, O1] =
     ZStream.absolve(ev(self))
 
+  /**
+   * Applies an aggregator to the stream, which converts one or more elements
+   * of type `A` into elements of type `B`.
+   */
+  def aggregate[R1 <: R, E1 >: E, O1 >: O, P](sink: ZTransducer[R1, E1, O1, P]): ZStream[R1, E1, P] =
+    ZStream {
+      for {
+        pull <- self.process
+        push <- sink.push
+        done <- ZRef.makeManaged(false)
+        run = {
+          def go: ZIO[R1, Option[E1], Chunk[P]] = done.get.flatMap {
+            if (_)
+              push(None).repeat(Schedule.doWhile(_.isEmpty))
+            else
+              pull
+                .foldM(
+                  _.fold(done.set(true) *> push(None))(Pull.fail(_)),
+                  os => push(Some(os))
+                )
+                .foldCauseM(
+                  Cause.sequenceCauseOption(_).fold(go)(Pull.halt(_)),
```

Comment:
The `fold(go)` here looks suspicious, because we can get here both from `push(Some(os))` and from `push(None)`.

- From `push(Some(os))` - this looks like a protocol violation, right? Do we allow transducers to end on their own accord? I'm not clear on this 🧐
- From `push(None)` - this will recurse again if the transducer ended in response to the stream ending. Ah, it'll end on the next iteration because of `done set true`. Ok, I'd prefer to `ZIO.fail(None)` directly here but not critical 👍 

So let's just clarify the first point. What do you think?

Reply:
I think this also happens with sinks, the way they signal end is by failing which, thanks to ZIO being well-typed, you have to account for everywhere. That's another instance of what we were trying to fix with `SignalFunction`. Not sure there is a solution other than writing laws and throwing all the transducers at them.

Reply:
I could refactor so that we die in case of a protocol error, but somehow that feels imperative. Note that having two different paths for the `None` and `Some` cases won't change the fact that the transducer may fail.

Reply:
So I was thinking about this for a bit and perhaps it’s useful to allow transducers to signal end? E.g. that allows to create transducers that can short-circuit a stream when a predicate is violated.

Reply:
Can you give an example? With `collectAllWhile`, I took the position that the transducer would just drop all elements violating the predicate, but still keep on accepting the rest of the stream. Is there another class of transducers where that wouldn't be possible?